### ISSUE TITLE
Failed to install SU2 by Python Wrapper Build

### ISSUE THREAD

**Issue by thw1021** [su2code/SU2#722]
Dear developers,
      I am trying to install SU2 v6.2.0 on Ubuntu 16.04 following the [instruction](https://su2code.github.io/docs/Python-Wrapper-Build/) you provided. But it failed with errors. I have installed Swig and mpi4py as you suggested.

The command I used:
```
./configure --prefix=/home/lab/SU2/SU2_Install CXXFLAGS="-O3" --enable-mpi --with-cc=/usr/bin/mpicc --with-cxx=/usr/bin/mpicxx --enable-PY_WRAPPER
```
The SU2_RUN and SU2_HOME environment variables displayed at the conclusion of configure.

When I run `make -j 24`, the error showed up:
```
make[2]: Entering directory '/home/lab/SU2/SU2_PY/pySU2'
swig -DHAVE_MPI  -Wall -I/home/lab/anaconda3/include/python3.7m -I/home/lab/anaconda3/include/python3.7m -I/home/lab/.local/lib/python2.7/site-packages/mpi4py/include -I/home/lab/anaconda3/lib/python3.7/site-packages/mpi4py/include -I/Library/Python/2.7/site-packages/mpi4py/include -outdir ./ -o SU2_APIPYTHON_wrap.cxx -c++ -python /home/lab/SU2/SU2_PY/pySU2/pySU2.i 
/home/lab/SU2/SU2_PY/pySU2/pySU2.i:64: Error: Unable to find 'mpi4py/mpi4py.i'
Makefile:532: recipe for target 'SU2_APIPYTHON_wrap.cxx' failed
make[2]: *** [SU2_APIPYTHON_wrap.cxx] Error 1
make[2]: Leaving directory '/home/lab/SU2/SU2_PY/pySU2'
Makefile:525: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/lab/SU2/SU2_PY/pySU2'
Makefile:441: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

Best.

**Comment by pcarruscag** [comment]
Do you have mpi4py installed in your machine?

**Comment by clarkpede** [comment]
I've run into a similar problem.  We can check if you're having the same issue (and figure out more about the problem) by doing the following.  Open a python terminal and enter the following commands:
```
from mpi4py import MPI
MPI
```
If you've got mpi4py installed on your system and it's on your python path, you'll see something like:
```
<module 'mpi4py.MPI' from '/home/clarkpede/.local/lib/python3.6/site-packages/mpi4py/MPI.cpython-34m.so'>
```
What output do you get from that command?

**Comment by thw1021** [comment]
Thank you. I have found my where my problem was. My default python environment is anaconda python. But when I use `sudo pip install mpi4py`, it actually installed mpi4py for python2.7 (system python). So I have to install mpi4py for python3 (system or anaconda python). Due to my network problem, I choose to intall mpi4py with following commands:
```
sudo apt install python3-pip
sudo pip3 install mpi4py
```
Now I can install SU2 by Python Wrapper Build without errors.

Thank you. Closing now.

**Comment by clarkpede** [comment]
For people who came across this issue later, my previous issue was very similar.  If you have multiple python environments installed on your machine (e.g. Python 2.7 and Python 3.6), then the build process for the python wrapper doesn't always select the one with mpi4py installed.  In my case, I had to manually edit the Makefile in `SU2_PY/pySU2/Makefile` to point to the correct python installation.

**Comment by thw1021** [comment]
@clarkpede 
I met a similiar problem here https://github.com/su2code/SU2/issues/739#issuecomment-515298427.

Could you show me how to modify the Makefile, please ?

**Comment by clarkpede** [comment]
It seems like your situation may have been resolved, but for archival purposes, I'll list my workaround here.

This problem occurs when you've installed python packages in two different locations.  This could be due to installing some python packages in a system directory (e.g. `/usr/lib/`) and some in a user directory (e.g. `~/.local/`).  SU2's build process is set to automatically detect the default package location, but it's not equipped to handle multiple package locations.  There are some hardcoded includes, but these only work for python 2.7.

I don't consider this a bug *per se* because multiple python package directories is not a standard use case.

#### The symptoms

When building SU2 with the python wrapper you'll see the following error during the swig build:
```
<install directory>/../SU2_PY/pySU2/pySU2.i:64: Error: Unable to find 'mpi4py/mpi4py.i'
```
But mpi4py is indeed installed on your system and it's on your python path, as confirmed by opening a python terminal and running `import mpi4py`.

#### The easiest workaround

Switch to a [virtualenv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/).  This will ensure that all of your python packages are stored in single location.

#### Alternate workaround

In this method, you change the hardcoded includes to match mpi4py's package location.

First I figure out where my mpi4py is installed.  I start a terminal session of python using the python environment I want to use (in this case, it's python 3.6), and then run:
```
from mpi4py import MPI
MPI
```

Since I have mpi4py installed on your system and it's on my python path, I see something like:
```
<module 'mpi4py.MPI' from '/home/clarkpede/.local/lib/python3.6/site-packages/mpi4py/MPI.cpython-34m.so'>
```

The `/home/clarkpede/.local/lib/python3.6/` directory is where my mpi4py package is installed.  But the swig compile command includes `/home/clarkpede/.local/lib/python2.7/`.  That's not the right directory! So first I find the hardcoded paths of the makefile by running (on a bash terminal):
```
grep -rn python2\.7 SU2_PY/ 
```
On my system, this prints out four lines. I don't care about the `Makefile.in` files, because those are generated automatically and will be overwritten every time I run "configure" or "preconfigure.py." The `NUMPY_INCLUDE` line is also commented out, so I ignore that too.  That leaves me with one line, line 51 of `SU2_PY/pySU2/Makefile.am`:

```
SU2_PY/pySU2/Makefile.am:51:MPI4PY_INCLUDE = ${HOME}/.local/lib/python2.7/site-packages/mpi4py/include \
```

I now modify line 51 of Makefile.am to read:

```
MPI4PY_INCLUDE = ${HOME}/.local/lib/python3.6/site-packages/mpi4py/include \
```

Then run configure or preconfigure.py again, and then run make again.  You should be good to go!

#### tl;dr

If you're having this error, modify the `MPI4PY_INCLUDE` line of `SU2_PY/pySU2/Makefile.am` to include the location of your mpi4py package.

**Comment by thw1021** [comment]
Thank you.

**Comment by ludgerpaehler** [comment]
If anyone is encountering this issue in 2022, and you have a virtualenv set up and the error appears: Do a user-install of mpi4py and it will be able to find it. I.e.

```bash
pip3 install mpi4py --user
```

**Comment by clarkpede** [comment]
Anyone coming across this issue in 2022: Makefiles are the "old way" to compile SU2.  Meson + ninja are the "new way" to compile SU2.  Editing makefiles will not change the meson + ninja build process.

Please, just use a virtualenv if your computer is struggling with different python versions/libraries.
